src/operator/tensor/indexing_op.h

```diff
@@ -317,23 +380,21 @@ void EmbeddingOpBackward(const nnvm::NodeAttrs& attrs,
 }
 
 template<int req>
-struct EmbeddingBackwardRsp {
+struct EmbeddingBackwardDns {
   template<typename DType, typename IType>
   // each thread i is responsible for target gradient row ids in [segment_start, segment_end)
-  MSHADOW_XINLINE static void Map(int i, const size_t width, IType* dst_idx, DType* dst_val,
+  MSHADOW_XINLINE static void Map(int i, const index_t num_cols, DType* dst_val,
                                   const IType* idx, const size_t num_idx, const DType* src,
-                                  const size_t segment_len, const size_t num_rows) {
+                                  const int segment_len, const size_t num_rows) {
     auto req_type = req;
-    size_t segment_start = i * segment_len;
-    size_t segment_end = (i + 1) * segment_len;
+    auto segment_start = i * segment_len;
+    auto segment_end = (i + 1) * segment_len;
     for (size_t y = 0; y < num_idx; y++) {
-      size_t j = idx[y];
+      auto j = idx[y];
       if (j >= num_rows) j = num_rows - 1;
       if (j < segment_start || j >= segment_end) continue;
-      dst_idx[j] = j;
-      for (size_t k = 0; k < width; k++) {
-        if (req_type == kWriteTo) req_type = kAddTo;
-        KERNEL_ASSIGN(dst_val[j * width + k], req_type, src[y * width + k]);
+      for (index_t k = 0; k < num_cols; k++) {
+        KERNEL_ASSIGN(dst_val[j * num_cols + k], req_type, src[y * num_cols + k]);
```

Comment:
Since `j*num_cols` and `y*num_cols` are not dependent on the loop, is it more efficient to calculate them outside the loop?